testing/src/main/java/com/linecorp/decaton/testing/processor/ProcessorTestSuite.java

```diff
@@ -0,0 +1,307 @@
+        }
+
+        public ProcessorTestSuite build() {
+            Set<ProcessingGuarantee> semantics = new HashSet<>();
+            for (GuaranteeType guaranteeType : defaultSemantics) {
+                semantics.add(guaranteeType.get());
+            }
+            semantics.addAll(customSemantics);
+
+            return new ProcessorTestSuite(rule,
+                                          configureProcessorsBuilder,
+                                          retryConfig,
+                                          propertySupplier,
+                                          semantics);
+        }
+    }
+
+    public static Builder builder(KafkaClusterRule rule) {
+        return new Builder(rule);
+    }
+
+    public void run() {
+        String topic = rule.admin().createRandomTopic(NUM_PARTITIONS, 3);
+        CountDownLatch rollingRestartLatch = new CountDownLatch(NUM_TASKS / 2);
+
+        DecatonProcessor<TestTask> preprocessor = (context, task) -> {
+            long startTime = System.nanoTime();
+            context.deferCompletion()
+                   .completeWith(context.push(task))
+                   .whenComplete((r, e) -> {
+                       semantics.forEach(
+                               g -> g.onProcess(new ProcessedRecord(context.key(),
+                                                                    task,
+                                                                    startTime,
+                                                                    System.nanoTime())));
+                   });
+            rollingRestartLatch.countDown();
+        };
+
+        Function<Integer, ProcessorSubscription> subscriptionConstructor = sequence -> {
+            ProcessorsBuilder<TestTask> processorsBuilder =
+                    configureProcessorsBuilder.apply(
+                            ProcessorsBuilder.consuming(topic, new TestTaskDeserializer())
+                                             .thenProcess(preprocessor));
+
+            return TestUtils.subscription("subscription-" + sequence,
+                                          rule.bootstrapServers(),
+                                          processorsBuilder,
+                                          retryConfig,
+                                          propertySuppliers);
+        };
+
+        Producer<String, DecatonTaskRequest> producer = null;
+        ProcessorSubscription[] subscriptions = new ProcessorSubscription[NUM_SUBSCRIPTION_INSTANCES];
+        try {
+            producer = TestUtils.producer(rule.bootstrapServers());
+            for (int i = 0; i < subscriptions.length; i++) {
+                subscriptions[i] = subscriptionConstructor.apply(i);
+            }
+            CompletableFuture<Map<Integer, Long>> produceFuture =
+                    produceTasks(producer, topic, record -> semantics.forEach(g -> g.onProduce(record)));
+
+            // perform rolling restart
+            rollingRestartLatch.await();
```

Comment:
Seems like this block can be much visible with some refactoring like extracting each step out into private methods.